discovery/hetzner/robot.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hetzner
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hetznerRobotLabelPrefix    = hetznerLabelPrefix + "robot_"
+	hetznerLabelRobotProduct   = hetznerRobotLabelPrefix + "product"
+	hetznerLabelRobotCancelled = hetznerRobotLabelPrefix + "cancelled"
+)
+
+// Discovery periodically performs Hetzner Robot requests. It implements
+// the Discoverer interface.
+type robotDiscovery struct {
+	*refresh.Discovery
+	client   *http.Client
+	port     int
+	endpoint string
+}
+
+// newRobotDiscovery returns a new robotDiscovery which periodically refreshes its targets.
+func newRobotDiscovery(conf *SDConfig, logger log.Logger) (*robotDiscovery, error) {
+	d := &robotDiscovery{
+		port:     conf.Port,
+		endpoint: conf.robotEndpoint,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+	d.client = &http.Client{
+		Transport: rt,
+		Timeout:   time.Duration(conf.RefreshInterval),
+	}
+
+	return d, nil
+}
+func (d *robotDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	resp, err := d.client.Get(d.endpoint + "/server")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, resp.Body)
```

Comment:
Why do you need this? The decoder should consume everything.

Reply:
We do this everywhere in the discover code base.

Reply:
Exhaust every request body before closing it (#5166) (#5479)

    From the documentation:
    > The default HTTP client's Transport may not
    > reuse HTTP/1.x "keep-alive" TCP connections if the Body is
    > not read to completion and closed.

    This effectively enable keep-alive for the fixed requests.

Reply:
@bradfitz doesn't mention the need to consume the entire body here in his boilerplate for the right way to make an HTTP request: https://github.com/bradfitz/exp-httpclient/blob/master/problems.md

If I'm reading the http code correctly, which I might not be... Ignoring cases where responses don't have bodies or when there's a protocol switch... It looks like the `readTransfer` [wraps an underlying reader](https://github.com/golang/go/blob/go1.15/src/net/http/transfer.go#L545-L565) with a `body` whose [Close method](https://github.com/golang/go/blob/go1.15/src/net/http/transfer.go#L961-L1000) reads until EOF (problem solved, right?)... But then `persistConn` [wraps that body in a `bodyEOFSignal`](https://github.com/golang/go/blob/go1.15/src/net/http/transport.go#L2121-L2141), whose job it is to keep track of reading until EOF but is incapable of observing that the `body` underneath is reading to EOF on Close (in the happy path)... So I think the http package is doing all the work of reading to EOF and blocking until it's done on `resp.Body.Close()`, but then it's closing the connection anyway.

Long story short, I think you need to manually read until EOF if you want to reuse a persistent connection in the event of a decoding failure (or extra whitespace at after the JSON object).